text_encoder.go

```diff
@@ -0,0 +1,226 @@
+// RFC3339-formatted timestamps.
+func NewTextEncoder(options ...TextOption) Encoder {
+	enc := textPool.Get().(*textEncoder)
+	enc.truncate()
+	enc.timeFmt = time.RFC3339
+	for _, opt := range options {
+		opt.apply(enc)
+	}
+	return enc
+}
+
+func (enc *textEncoder) Free() {
+	textPool.Put(enc)
+}
+
+func (enc *textEncoder) AddString(key, val string) {
+	enc.addKey(key)
+	enc.bytes = append(enc.bytes, val...)
+}
+
+func (enc *textEncoder) AddBool(key string, val bool) {
+	enc.addKey(key)
+	if val {
+		enc.bytes = append(enc.bytes, "true"...)
+		return
+	}
+	enc.bytes = append(enc.bytes, "false"...)
+}
+
+func (enc *textEncoder) AddInt(key string, val int) {
+	enc.AddInt64(key, int64(val))
+}
+
+func (enc *textEncoder) AddInt64(key string, val int64) {
+	enc.addKey(key)
+	enc.bytes = strconv.AppendInt(enc.bytes, val, 10)
+}
+
+// TODO: uncomment once #116 lands.
+// func (enc *textEncoder) AddUint(key string, val uint) {
+// 	enc.AddUint64(key, uint64(val))
+// }
+
+// func (enc *textEncoder) AddUint64(key string, val uint64) {
+// 	enc.addKey(key)
+// 	enc.bytes = strconv.AppendUint(enc.bytes, val, 10)
+// }
+
+func (enc *textEncoder) AddFloat64(key string, val float64) {
+	enc.addKey(key)
+	switch {
+	case math.IsNaN(val):
+		enc.bytes = append(enc.bytes, "NaN"...)
+	case math.IsInf(val, 1):
+		enc.bytes = append(enc.bytes, "+Inf"...)
+	case math.IsInf(val, -1):
+		enc.bytes = append(enc.bytes, "-Inf"...)
+	default:
+		enc.bytes = strconv.AppendFloat(enc.bytes, val, 'f', -1, 64)
+	}
+}
+
+func (enc *textEncoder) AddMarshaler(key string, obj LogMarshaler) error {
+	enc.addKey(key)
+	enc.bytes = append(enc.bytes, '{')
+	err := obj.MarshalLog(enc)
+	enc.bytes = append(enc.bytes, '}')
+	return err
+}
+
+func (enc *textEncoder) AddObject(key string, obj interface{}) error {
+	enc.AddString(key, fmt.Sprintf("%+v", obj))
+	return nil
+}
+
+func (enc *textEncoder) Clone() Encoder {
+	clone := textPool.Get().(*textEncoder)
+	clone.truncate()
+	clone.bytes = append(clone.bytes, enc.bytes...)
+	clone.timeFmt = enc.timeFmt
+	return clone
+}
+
+func (enc *textEncoder) WriteEntry(sink io.Writer, msg string, lvl Level, t time.Time) error {
+	if sink == nil {
+		return errNilSink
+	}
+
+	final := textPool.Get().(*textEncoder)
+	final.truncate()
+	enc.addLevel(final, lvl)
+	enc.addTime(final, t)
+	enc.addMessage(final, msg)
+
+	if len(enc.bytes) > 0 {
+		final.bytes = append(final.bytes, ' ')
+		final.bytes = append(final.bytes, enc.bytes...)
+	}
+	final.bytes = append(final.bytes, '\n')
+
+	expectedBytes := len(final.bytes)
+	n, err := sink.Write(final.bytes)
+	final.Free()
+	if err != nil {
+		return err
+	}
+	if n != expectedBytes {
+		return fmt.Errorf("incomplete write: only wrote %v of %v bytes", n, expectedBytes)
+	}
+	return nil
+}
+
+func (enc *textEncoder) truncate() {
+	enc.bytes = enc.bytes[:0]
+}
+
+func (enc *textEncoder) addKey(key string) {
+	lastIdx := len(enc.bytes) - 1
+	if lastIdx >= 0 && enc.bytes[lastIdx] != '{' {
```

Comment:
Sure,
```go
func ExampleNew_textEncoderBug() {
	textLogger := New(NewTextEncoder(
		TextNoTime(), // drop timestamps in tests.
	))
	textLogger.Info("Test 1.")
	textLogger.Info("With fields", String("f1", "v1"), String("f2", "v2"))
	textLogger.Info("With fields", String("f1", "{"), String("f2", "v2"))

	// Output:
	// [I] Test 1.
	// [I] With Fields f1=v1 f2=v2
	// [I] With Fields f1={ f2=v2
}
```

This currently fails:
```
got:
[I] Test 1.
[I] With fields f1=v1 f2=v2
[I] With fields f1={f2=v2
want:
[I] Test 1.
[I] With Fields f1=v1 f2=v2
[I] With Fields f1={ f2=v2
```

Basically the `addKey` is supposed to add ` key=` (space at the beginning) for keys added to the top level object.

However, since we also support adding nested objects, if the user adds something using a log marshaler, we want the output to look something like:
```
[I] Log message nested={a=1 b=2}
```

To get that to work, the code currently looks at the last character, and if it's `{`, it drops the space in front. However, this assumes that the user's values for other fields cannot end in a `{`, which I'd prefer not to assume.

To fix this properly, it needs the encoder to keep some state for when it's inside of an object (and we're at the first field) to avoid the space.